File: Ruby/DeepSpaceRuby/lib/Hangar.rb
#encoding:utf-8

require_relative 'HangarToUI.rb'

module Deepspace
  class Hangar
    attr_reader :maxElements, :weapons, :shieldBoosters

    def initialize (m)
      @maxElements = m
      @weapons = Array.new
      @shieldBoosters = Array.new
    end

    def self.newCopy (h)
      copy = Hangar.new(h.maxElements)

      h.weapons.each { |w|
        copy.weapons.push(Weapon.newCopy(w))
      }

      h.shieldBoosters.each { |s|
        copy.shieldBoosters.push(ShieldBooster.newCopy(s))
      }
      return copy
    end

    def getUIversion
      return HangarToUI.new(self)
    end

    private
    def spaceAvailable
      if ((@weapons.length + @shieldBoosters.length) < @maxElements) then
        return true
      else
        return false
      end
    end

    public
    def addWeapon (w)
      if spaceAvailable then
        @weapons.push(w);
        return true;
      else
        return false;
      end
    end

    def addShieldBooster (s)
      if spaceAvailable then
        @shieldBoosters.push(s)
      end
    end

    def removeShieldBooster (s)
      if (s >= @shieldBoosters.length || s < 0) then
        return nil
      else
        return @shieldBoosters.delete_at(s)
      end
    end

    def removeWeapon (w)
      if (w >= @weapons.length || w < 0) then
        return nil
      else
        return @weapons.delete_at(w)
      end
    end

    def to_s
      cad = "maxElements: #{@maxElements}"
      cad += "\nArmas: \n"

      @weapons.each{ |w|
        cad += w.to_s
        cad += "\n"
      }

      cad += "\nShieldBoosters: \n"

      @shieldBoosters.each{ |s|
        cad += s.to_s
        cad += "\n"
      }

      return cad
    end

  end
end

File: Java/DeepspaceJava/deepspace/src/deepspace/Hangar.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author paula
 */
public class Hangar {
    private int maxElements;
    private ArrayList<Weapon> weapons;
    private ArrayList<ShieldBooster> shieldBoosters;

    Hangar (int capacity){
        maxElements = capacity;
        weapons = new ArrayList<>();
        shieldBoosters = new ArrayList<>();
        shieldBoosters = new ArrayList<>();
    }

    Hangar (Hangar h){
        maxElements = h.getMaxElements();
        weapons = new ArrayList<>();
        weapons = h.getWeapons();
        shieldBoosters = new ArrayList<>();
        shieldBoosters = h.getShieldBoosters();
    }
            
    public int getMaxElements (){
        return maxElements;
    }
    
    public ArrayList<Weapon> getWeapons (){
        return weapons;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters (){
        return shieldBoosters;
    }
    
    public HangarToUI getUIversion (){
      return new HangarToUI(this);
    }

    private boolean spaceAvailable (){
        if ((weapons.size() + shieldBoosters.size()) < maxElements)
            return true;
        else
            return false;
    }

    public boolean addWeapon (Weapon w){
        if (spaceAvailable()){
            weapons.add(w);
            return true;
        }
        else
            return false;
    }

    public boolean addShieldBooster (ShieldBooster s){
        if (spaceAvailable()){
            shieldBoosters.add(s);
            return true;
        }
        else
            return false;
    }

    public ShieldBooster removeShieldBooster (int s){
        if (s>=shieldBoosters.size())
            return null;
        else{
            return shieldBoosters.remove(s);
        }
    }
    
    public Weapon removeWeapon (int w){
        if (w>=weapons.size())
            return null;
        else{
            return weapons.remove(w);
        }
    }
    
    @Override
    public String toString(){
        String res="Armas:\n "+weapons.toString()+
                "  \nPotenciadores de escudo:\n"+shieldBoosters.toString()+
                "\nElementos maximos: "+maxElements;
        return res;
    }
}
